src/PackageCompiler.jl

```diff
@@ -1200,11 +1200,23 @@ function bundle_julia_libraries(dest_dir, stdlibs)
     tot_libsize = 0
     printstyled("PackageCompiler: bundled libraries:\n")
 
-    # Reqiored libraries
+    # Bundle the libstdc++ that is actually loaded by Julia
+    # xref: https://discourse.julialang.org/t/precedence-of-local-and-julia-shipped-shared-libraries/104258?u=sloede
+    libstdcpp_search = filter(contains("libstdc++"), Libdl.dllist())
+    if isempty(libstdcpp_search)
+        # Fall back to default search directory if libstdc++ is not loaded
+        libstdcpp_dir = libjulia_dir
+    else
+        # If libstdc++ was found, get absolute path to directory
+        libstdcpp_dir = dirname(abspath(first(libstdcpp_search)))
+    end
```

Comment:
Another option is to just `dlopen("libstdc++")` before running `Libdl.dllist()`:

On macOS that will load the one that comes with Julia:
```
julia> using Libdl
       Libdl.dllist()
       @show filter(contains("libstdc++"), Libdl.dllist())
       dlopen("libstdc++")
       @show filter(contains("libstdc++"), Libdl.dllist());
filter(contains("libstdc++"), Libdl.dllist()) = String[]
filter(contains("libstdc++"), Libdl.dllist()) = ["/Users/sabae/.julia/juliaup/julia-1.9.3+0.aarch64.apple.darwin14/lib/julia/libstdc++.6.dylib"]
```

On Linux (where we do the fancy `libstdc++` search) you'll already have `libstdc++` loaded, so the `dlopen("libstdc++")` doesn't do anything.  The only reason I suggest this is that then `Libdl.dllist()` will always then give you the absolute path to the file, which may be less prone to edge cases than searchinmg in a directory.

Reply:
You think it could happen that `Libdl.dllist()` contains a library without a full path?

The reason I implemented it similar to the existing search is that this way you end up with, e.g.,
```
lrwxrwxrwx 1 mschlott mi      19 Sep 27 09:17 libstdc++.so.6 -> libstdc++.so.6.0.28
-rw-r--r-- 1 mschlott mi 1956992 Sep 27 09:17 libstdc++.so.6.0.28
```
in `PREFIX/lib/julia` for a `/usr/lib/x86_64-linux-gnu` directory with
```
lrwxrwxrwx 1 root root      19 Jul  9 08:45 /usr/lib/x86_64-linux-gnu/libstdc++.so.6 -> libstdc++.so.6.0.28
-rw-r--r-- 1 root root 1956992 Jul  9 08:45 /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.28
```
Otherwise, you'd have to `cp(src, dst; follow_symlinks=true)` to make sure that symbolic links are resolved before copying.

But if you think that's better/saner/more robust, I'm happy to change it as you suggested.

Reply:
The reason for `libstdc++.so.6 -> libstdc++.so.6.0.28` to exist is so that a wide variety of applications can link against `libstdc++.so.6` and find the file on-disk, but once it's already in memory that shouldn't matter, it should look at the `SONAME` of the loaded library and short-circuit based on that.  I like using `Libdl.dllist()`'s output (which should always be an absolute pathname to the actual file) to avoid problems like finding a linker script instead of the actual `.so`, bundling multiple versions of the same library if you somehow have multiple, etc....  If you need a library, I just think it's simpler to copy precisely the file that ends up getting loaded.

Reply:
I tried your suggestion, but I ran into new troubles. On my test system, the system libstdc++ is actually older than the one bundled with Julia. Thus, before I used
```shell
LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.28 my/packagecompiler/build/command
```
to make sure that I can test with using a non-Julia libstdc++. In this case, however, Julia will actually load *both* libraries, and
```julia
dlopen("libstdc++")
@show filter(contains("libstdc++"), Libdl.dllist());
```
yields
```
["/usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.28", "/mnt/hd1/opt/julia/1.9.3/bin/../lib/julia/libstdc++.so"]
```

Thus it seems if someone is intent on preloading libstdc++ (which is not unheard of), this new approach will fail - indeed, `dlopen` *will* load a second library if they do not match 100%.

I then tried the following, where I tested whether libstdc++ is already loaded and then call `Libdl.dlopen` only if it isn't. This remedies the problem before, but now a second issue arises:

While `dllist()` contains absolute paths, it does _not_ resolve symbolic links. Since Julia looks for `libstdc++.so.6`, it will stop and use the first it finds, e.g., `/mnt/hd1/opt/julia/1.9.3/bin/../lib/julia/libstdc++.so.6`. However, I cannot just copy this path, since it is just a symbolic link. If I call `realpath` before copying, this will, however, put `libstdc++.so.6.0.30` in my bundle, which will _not_ be picked up when Julia searches for `libstdc++.so.6` (I verified this using `LD_DEBUG=libs`). Thus, the next option would be that I copy the library _without_ resolving symlinks, but use `follow_symlinks=true`. This will work well, but now we end up with only `libstdc++.so.6` in the bundle without being able to tell which version it actually is. It also seems fishy to me to leave it at that, given that is _not_ how the symlinks are organized on Linux/macOS systems.

Therefore, what I seem to end up with (as far as I can tell), is the following:
* Use `dlopen` to load `libstdc++` if not yet present in `dllist()`
* Get path to `libstdc++` from `dllist()`
* Expand path to `realpath(abspath(...))`
* Copy this `libstdc++` to `PREFIX/lib/julia` (e.g., `libstdc++.so.6.0.28` or `libstdc++.6.dylib` or `libstdc++.dll`)
* Create relative symlinks within `PREFIX/lib/julia`:
  * `Sys.iswindows()`: nothing to do
  * `Sys.isapple()`:
    ```
    libstdc++.dylib -> libstdc++.6.dylib
    ```
  * `Sys.isunix()`:
    ```
    libstdc++.so -> libstdc++.so.6.0.28
    libstdc++.so.6 -> libstdc++.so.6.0.28
    ```

This seems to be very involved 😅, thus it would be great to get some feedback whether this sounds correct or if I'm missing something else.

Reply:
I went ahead and implemented this newest approach in https://github.com/JuliaLang/PackageCompiler.jl/pull/853/commits/21c2f726ccd78fe95f35df2d954b864bf8020330. I am having it tested it by the colleague who filed the original issue, and we will report back.

Reply:
Hi there!

I just confirmed that  @sloede 's current implementation fixes my initial issue, i.e. with his fix my system `libstdc++.so.6.0.32` is correctly picked up by PackageCompiler.